src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -293,8 +368,69 @@ private TextSpan GetRegionSpanForReference(SourceText sourceText, TextSpan refer
                     sourceText.Lines[lastLineNumber].End);
             }
 
+            private void UpdateCustomColumnsVisibility(MultiDictionary<string, string> customData)
+            {
+                // Check if we have any custom reference data to display.
+                if (customData.Count == 0)
+                {
+                    return;
+                }
+
+                // Get the new column states corresponding to the custom columns to display for custom data.
+                List<ColumnState2> newColumnStatesOpt = null;
+
+                lock (Gate)
+                {
+                    foreach (var customColumnName in customData.Keys)
+                    {
+                        // Get the matching custom column.
+                        var customColumnDefinition = TableControl.ColumnDefinitionManager.GetColumnDefinition(customColumnName) as AbstractFindUsagesCustomColumnDefinition;
+                        if (customColumnDefinition == null)
+                        {
+                            Debug.Fail($"{nameof(SourceReferenceItem.ReferenceInfo)} has a key '{customColumnName}', but there is no exported '{nameof(AbstractFindUsagesCustomColumnDefinition)}' with this name.");
+                            continue;
+                        }
+
+                        // Ensure that we flip the visibility to true for the custom column.
+                        ColumnState2 newColumnStateOpt = null;
+                        if (_customColumnTitleToStatesMap.TryGetValue(customColumnDefinition.Name, out var columnState))
+                        {
+                            if (!columnState.IsVisible)
+                            {
+                                newColumnStateOpt = new ColumnState2(columnState.Name, isVisible: true, columnState.Width,
+                                    columnState.SortPriority, columnState.DescendingSort, columnState.GroupingPriority);
```

Comment:
just to verify, but this is just a simple object. t his isn't doing something UI/costly in the lock right?

Reply:
Yep, this is just a state object to communicate about column state changes to editor layer in subsequent calls. So this creation is cheap, the expensive/UI operation is invocation of `TableControl.SetColumnStates`